java/src/Ice/src/main/java/com/zeroc/IceInternal/Instance.java

```diff
@@ -720,90 +720,82 @@ public Class<?> apply(String typeId)
-        if(fullyQualifiedClassName == null)
+        if (fullyQualifiedClassName != null)
         {
-            fullyQualifiedClassName = com.zeroc.Ice.Util.typeIdToClass(typeId);
-            addClass = true;
+            return getConcreteClass(fullyQualifiedClassName);
         }
 
         //
-        // See if we can find the class without any prefix.
+        // It's a new type ID, so first convert it into a Java class name.
         //
-        c = getConcreteClass(fullyQualifiedClassName);
+        fullyQualifiedClassName = com.zeroc.Ice.Util.typeIdToClass(typeId);
 
         //
-        // See if the application defined an Ice.Package.MODULE property.
+        // See if the application defined any package prefixes with Ice.Package.MODULE property.
         //
-        if(c == null)
+        int pos = typeId.indexOf(':', 2);
+        if (pos != -1)
         {
-            int pos = typeId.indexOf(':', 2);
-            if(pos != -1)
+            String topLevelModule = typeId.substring(2, pos);
+            String[] packagePrefixes = _builtInModulePackagePrefixes.get(topLevelModule);
+            if (packagePrefixes == null)
             {
-                String topLevelModule = typeId.substring(2, pos);
-                String[] packagePrefixes = _builtInModulePackagePrefixes.get(topLevelModule);
-                if (packagePrefixes == null)
-                {
-                    packagePrefixes = _initData.properties.getPropertyAsListWithDefault("Ice.Package." + topLevelModule, null);
-                }
+                packagePrefixes = _initData.properties.getPropertyAsListWithDefault("Ice.Package." + topLevelModule, null);
+            }
 
-                if(packagePrefixes != null)
+            if (packagePrefixes != null)
+            {
+                for (String packagePrefix : packagePrefixes)
                 {
-                    for(String packagePrefix : packagePrefixes)
+                    c = getConcreteClass(packagePrefix + "." + fullyQualifiedClassName);
+                    if (c != null)
                     {
-                        c = getConcreteClass(packagePrefix + "." + fullyQualifiedClassName);
-                        if(c != null)
-                        {
-                            break;
-                        }
+                        break;
                     }
                 }
             }
         }
 
-        //
-        // See if the application defined a default package.
-        //
-        if(c == null)
+        // If we didn't find the class yet, try the default package prefix or without prefix.
+        if (c == null)
         {
-            String pkg = _initData.properties.getProperty("Ice.Default.Package");
-            if(pkg.length() > 0)
-            {
-                c = getConcreteClass(pkg + "." + fullyQualifiedClassName);
-            }
+            String packagePrefix = _initData.properties.getProperty("Ice.Default.Package");
+            c = getConcreteClass(
```

Comment:
load the class by first trying to prepend the default package prefix to the fully qualified name returned by typeIdToClass. If no default package prefix is defined (i.e., packagePrefix is empty), we then directly use the fully qualified name as returned by typeIdToClass